drivers/rtc/rtc_am1805.c

```diff
@@ -0,0 +1,601 @@
+#include <zephyr/sys/util.h>
+
+#define DT_DRV_COMPAT ambiq_am1805
+
+LOG_MODULE_REGISTER(am1805, CONFIG_RTC_LOG_LEVEL);
+
+#define AM1805_IDENTITY_CODE	0x69
+
+/* AM1805 register address */
+#define REG_HUNDREDS_ADDR	0x00
+#define REG_SECONDS_ADDR	0x01
+#define REG_MINUTES_ADDR	0x02
+#define REG_HOURS_ADDR		0x03
+#define REG_MDAY_ADDR		0x04
+#define REG_MONTH_ADDR		0x05
+#define REG_YEAR_ADDR		0x06
+#define REG_WDAY_ADDR		0x07
+
+#define REG_ALM_HUNDREDS_ADDR	0x08
+#define REG_ALM_SECONDS_ADDR	0x09
+#define REG_ALM_MINUTES_ADDR	0x0A
+#define REG_ALM_HOURS_ADDR	0x0B
+#define REG_ALM_MDAY_ADDR	0x0C
+#define REG_ALM_MONTH_ADDR	0x0D
+#define REG_ALM_WDAY_ADDR	0x0E
+#define REG_STATUS_ADDR		0x0F
+#define REG_CONTROL1_ADDR	0x10
+#define REG_CONTROL2_ADDR	0x11
+#define REG_XT_CALIB_ADDR	0x14
+#define REG_TIMER_CTRL_ADDR	0x18
+#define REG_IRQ_MASK_ADDR       0x12
+#define REG_WATCHDOG_ADDR	0x1B
+#define REG_OSC_STATUS_ADDR	0x1D
+
+/* AM1805 control bits */
+#define SECONDS_BITS		GENMASK(6, 0)
+#define MINUTES_BITS		GENMASK(6, 0)
+#define HOURS_BITS		GENMASK(5, 0)
+#define DATE_BITS		GENMASK(5, 0)
+#define MONTHS_BITS		GENMASK(4, 0)
+#define WEEKDAY_BITS		GENMASK(2, 0)
+#define YEAR_BITS		GENMASK(7, 0)
+#define REG_CONTROL2_OUT2S_BITS GENMASK(4, 2)
+#define REG_TIMER_CTRL_RPT_BITS GENMASK(4, 2)
+#define REG_XT_CALIB_OFF_MASK	GENMASK(6, 0)
+
+#define REG_STATUS_ALM		BIT(2)
+#define REG_CONTROL1_STOP	BIT(7)
+#define REG_IRQ_MASK_AIE        BIT(2)
+#define REG_XT_CALIB_CMDX       BIT(7)
+
+#define TIMER_CTRL_ALM_OFF	0x00
+#define TIMER_CTRL_ALM_DAY	BIT(4)
+#define TIMER_CTRL_ALM_YEAR	BIT(2)
+#define TIMER_CTRL_ALM_HR	(BIT(2) | BIT(4))
+#define TIMER_CTRL_ALM_SEC	GENMASK(4, 2)
+#define TIMER_CTRL_ALM_MIN	GENMASK(4, 3)
+#define TIMER_CTRL_ALM_WEEK	GENMASK(3, 2)
+
+#define REG_WATCHDOG_WDS	BIT(7)
+#define WRB_1_SECOND		BIT(1)
+#define WRB_4_SECONDS		GENMASK(1, 0)
+
+#define REG_OSC_STATUS_ACAL_0	0x00
+#define REG_OSC_STATUS_ACAL_1	BIT(6)
+#define REG_OSC_STATUS_ACAL_2	BIT(7)
+#define REG_OSC_STATUS_ACAL_3	GENMASK(7, 6)
+#define REG_STATUS_DEFAULT	0x00
+
+#define AM1805_RTC_ALARM_TIME_MASK                                                                 \
+	(RTC_ALARM_TIME_MASK_SECOND | RTC_ALARM_TIME_MASK_MINUTE | RTC_ALARM_TIME_MASK_HOUR |      \
+	 RTC_ALARM_TIME_MASK_MONTHDAY | RTC_ALARM_TIME_MASK_MONTH | RTC_ALARM_TIME_MASK_WEEKDAY)
+
+#ifdef CONFIG_RTC_ALARM
+/* am1805-gpios property must be in the devicetree inorder to use the RTC_ALARM */
+#if !DT_ANY_INST_HAS_PROP_STATUS_OKAY(am1805_gpios)
+#error "am1805-gpios" - property not available in devicetree.
+#endif
+#endif
+
+struct am1805_config {
+	const struct i2c_dt_spec int_i2c;
+#ifdef CONFIG_RTC_ALARM
+	struct gpio_dt_spec int_gpio;
+#endif
+};
+
+struct am1805_data {
+	struct k_mutex lock;
+#ifdef CONFIG_RTC_ALARM
+	rtc_alarm_callback alarm_user_callback;
+	void *alarm_user_data;
+	/* For gpio-interrupt */
+	struct gpio_callback am1805_callback;
+	struct k_thread am1805_thread;
```

Comment:
> I see that's what `rtc_pcf8523.c` does, but `rtc_pcf8563.c` fires up the callback in the worqueue, which feels like a more reasonable approach (the application could always handle have separate thread if needed).You'd normally want a thread if you know that you want to sleep in it.


@fabiobaltieri  IMO,when it is fired up in the work-queue , It easily becomes less sure how precise it is going to be called(even when using separate thread in application, it is still served  in the work-queue from driver) as work-queue load is uncertain and sometimes very high.


> Seems like a bit of a waste of resources


For a more precise call and not to worry about the unsure load in the work-queue,IMO, 512 as the size of thread ,I feel its okay to use a separate thread than waiting in the queue.

Reply:
> @fabiobaltieri IMO,when it is fired up in the work-queue , It easily becomes less sure how precise it is going to be called(even when using separate thread in application, it is still served in the work-queue from driver) as work-queue load is uncertain and sometimes very high.

Sure, but here we are talking about a device where you set an alarm in *seconds* in the first place, does it make a difference if the handler is sitting in the workqueue for some tens of milliseconds? (if it's more than that something is abusing the workqueue and should be fixed).

Up to the maintainer for what I'm concerned but the problem is that if you implement the handler as a thread there's no way out, you'll have to allocate a stack for it (and SRAM is usually the bottleneck), while if you handle it on the workqueue then the user can do whatever they want with it. My line of thought is that unless the handler needs to sleep in the processing code for whatever reason, callback handlers should be implemented in a workqueue.

Regardless, not a big deal, don't consider it a blocking comment, but it'd be nice to avoid the story we have with the sensor drivers where half of them have different options for the trigger handler (see the various `_OWN_THREAD` Kconfigs) and it's more or less the same code copied all over.